Library Name

KevinCathcart opened the pull request:
Sorry for the delay in getting this put together.

This pull request incorporates two different approaches. The two different commits show each of the approaches.

The first approach is simply a string parameter passed to `singleSpaAngularWebpack`. To be backwards compatible, it falls back to trying to get the project name from angular.json. If that fails, it uses a default value, and prints a warning message that explains why a default value might be undesirable, and a link to information on how to fix it.

The second approach is similar, but uses a property defined in angular.json. It has equivalent fallback to the first approach. For this second approach, I've also shown an example of another property (`libraryTarget`) that might be helpful, especially since that was supported back pre-Angular 8 builder.

Once we finalize the approach we want to take, and get this into mergeable shape for the 4.x branch series, I'll backport it to 3.x. (See the 3.xLibraryName branch in my fork if desired).

## PR Checklist

Please check if your PR fulfills the following requirements:

- [ ] Tests for the changes have been added (for bug fixes / features)

Not clear what the best way to even approach automated testing would be. Trying to parse the webpack output seems like a bad idea. I suppose technically setting something up to try to pull it in via named amd could work, but that feels obnoxious at best.

## PR Type

What kind of change does this PR introduce?

- [x] Feature

## What is the current behavior?

Issue Number: #220 
## Does this PR introduce a breaking change?

- [x] No

KevinCathcart commented on src/webpack/index.ts:
```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
+  }
+  console.info('See <SomeUrl> for information on how to resolve this.');
+
+  return 'angular_single_spa_project';
```
I wanted something better than 'app3' as the default. I'm certainly willing to accept suggestions for a better default.

KevinCathcart commented on src/webpack/index.ts:
```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
+  console.warn(
+    'Warning: single-spa-angular could not determine a library name to use and has used a default value.',
+  );
+  console.info('This may cause issues if this app uses code-splitting or lazy loading.');
```
Using .info() for the follow up lines instead of .warn(), because multiple lines of bold yellow only makes things hard to read.

KevinCathcart commented on src/webpack/index.ts:
```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
+  if (!options) {
+    console.info('You may also need to update extra-webpack.config.json.');
+  }
+  console.info('See <SomeUrl> for information on how to resolve this.');
```
I think that linking to an explanation would work better than trying to include it in the console output.
Of course, this explanation still needs to be written, and the url determined.

joeldenning replied:
Perhaps we could add some documentation to the Angular page at https://github.com/single-spa/single-spa.js.org/blob/master/website/versioned_docs/version-5.x/ecosystem-angular.md?

KevinCathcart replied:
Agreed. That seemed like the most reasonable place to put it.

KevinCathcart replied:
I was thinking a new section before "Use Angular Builder" under manual installation could briefly document manually setting up the custom-webpack builder and document the new options, and we could point the users to that section.

joeldenning replied:
I think the Use Angular Builder section of the docs refers to the Angular 7 builder, not to the custom webpack builder. I think that this could be near the section about what the schematics do?

KevinCathcart commented on schematics/ng-add/index.ts:
```diff
@@ -140,6 +146,8 @@ function updateProjectNewAngular(context: SchematicContext, clientProject: Works
   (buildTarget.options as CustomWebpackBuilderOptions).customWebpackConfig = {
     path: join(clientProject.root, 'extra-webpack.config.js'),
+    libraryName: projectName,
+    libraryTarget: 'umd',
```
By changing this line, it is possible to change the default libraryTarget (perhaps to system), without it being a breaking change, as previously created apps would still use fallback default in webpack/index.ts.

joeldenning commented:
@arturovt do you have time to review this one?

joeldenning commented on src/webpack/index.ts:
```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
+  const angularJson = require(angularJsonPath);
+  if (!angularJson.projects) return null;
+
+  var projects = Object.getOwnPropertyNames(angularJson.projects);
```
Wouldn't `Object.keys()` work fine here? JSON files don't have non-enumerable properties, nor do they have prototype inheritance, so I don't think we'd lose anything by using the more common and familiar Object.keys function.

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/keys

KevinCathcart replied:
Yeah `Object.keys` should be fine. I normally reach for `getOwnPropertyNames` out of habit since it is usually the safest option unless I need prototypical inheritance.

joeldenning commented on src/webpack/index.ts:
```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
+  let current = process.cwd();
+  const root = path.parse(current).root;
+
+  while (current !== root) {
```
This looks like an infinite loop - we are never updating `current` inside of the loop.

KevinCathcart replied:
Yeah, the while loop is broken. I must have accidentally lost a line.

KevinCathcart replied:
Fixed in latest push.

arturovt commented on src/webpack/index.ts:
```diff
@@ -75,3 +84,55 @@ function removeMiniCssExtract(config: any) {
+  return null;
+}
+
+function findUp(files: string[]): string | null {
```
From my subjective perspective, this seems like a complicated implementation.

We have to be using tools that Angular DevKit already provides and you can work with Angular config using native tools:
```js
const { workspaces } = require('@angular-devkit/core');
const { NodeJsSyncHost } = require('@angular-devkit/core/node');

const host = workspaces.createWorkspaceHost(new NodeJsSyncHost());

workspaces.readWorkspace(`${process.cwd()}/angular.json`, host).then(({ workspace }) => {
  console.log(workspace.projects.keys());
});
```

It's also worth adding comments to the code. I've read it multiple times and I can't understand what's going on in `while + nested for` :smile:
Excuse my pedantry :)

KevinCathcart replied:
Yeah, the while loop is broken. I must have accidentally lost a line.

The problem I am trying to tackle here is that there is zero guarantee that angular.json is in the same folder as `process.cwd()`. This can happen if you run `ng build` from inside a source subfolder.
In order to find the workspace folder, upward searching is required.

All that is required before you can even call `.readWorkspace()`. 

angular-cli does its equivalent findUp logic here: https://github.com/angular/angular-cli/blob/0070ea46e6eaa25ad49d17c24fe7bed546dc1ec0/packages/angular/cli/utilities/project.ts#L29.

Once we have located the file, then yeah, we *could* use the devkit workspaces, except this is 100% synchronous code, and dev-kit workspaces is async.

arturovt replied:
Webpack config can export a Promise or a factory that returns Promise thus you're able to work with sync/async code.

KevinCathcart replied:
Yeah, but that defeats the whole point of the angular.json parsing, which is as a fallback that just works for the majority of all users if they are upgrading from an older version of the library. 

If the users need to update the `extra-webpack.config.ts` to make it export an promise returning function, then they could instead update it to pass the options argument to `singleSpaAngularWebpack` again, and update angular.json to specify the new 'libraryName' option. The new option (automatically populated by the schematic on new projects) is the primary source of the library name, with the angular.json parsing fallback as secondary, and a static value + warning as the third.

We could skip all angular.json parsing, but if we did so, then all existing users would either start getting the warning about using the default, and would need to make changes to silence it. The current approach is that only those few users who have multi-project workspaces will get the warning in the first place, as for everybody else we can infer it from the one-and-only project in angular.json.

arturovt replied:
Okay. Can't we re-use the existing function then and not copy it from their source code?:
```js
const { findUp } = require('@angular/cli/utilities/find-up');
```

KevinCathcart replied:
It was a from scratch re-write, not a copy.

But yeah, we could do that, with the caveat that it is not documented as a public api, so could potentially break in the future. (On the other hand custom builders were not documented as a public API before angular 8, and that was not a considered a problem for this package for angular 7.)

I'm perfectly happy to switch to that if that's what you want to do.

joeldenning replied:
I vote we switch over to `@angular/cli/utilities/find-up`

KevinCathcart commented:
Fixed several of the mentioned things.

Also switched the `require(angularJsonPath)` to `JSON.parse(fs.readFileSync(angularJsonPath, 'utf8'))` since that is more explicit about what we are doing, and avoids a lint suppression.

joeldenning commented:
Any update on this? Switching to `@angular/cli/utilities/find-up` seems to be the last item to be addressed?

KevinCathcart commented:
Yeah. Sorry, I've gotten a bit busy, and have not had a chance to make that
change and push it. I'll try to make time for it this evening.

On Wed, Jul 1, 2020, 12:21 Joel Denning <notifications@github.com> wrote:

> Any update on this? Switching to @angular/cli/utilities/find-up seems to
> be the last item to be addressed?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/single-spa/single-spa-angular/pull/229#issuecomment-652517307>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWEZJC65CCJ2PNKM27CE3RZNO7FANCNFSM4OD7OLNA>
> .
>

KevinCathcart commented:
I've pushed up a simplified version that uses '@angular/cli/utilities/find-up'.

The only reason I've still left this marked as a draft is that it links to a part of the documentation that does not yet exist:
https://single-spa.js.org/docs/ecosystem-angular/#use-custom-webpack

This would be a new section similar to the existing "Use Angular Builder" that documents these two options from angular.json, and explains how to make sure you are passing the options through in `custom-webpack.config.js`, probably by simply showing these two lines from the template:
```
module.exports = (config, options) => {
  const singleSpaWebpackConfig = singleSpaAngularWebpack(config, options);
```
Or even quoting the whole template since it is only 8 lines long.

I won't have a chance to draft up that documentation section tonight.

arturovt commented:
Hey @KevinCathcart 
Can I ask you not to squash commits? It's really hard to review what has been changed since the last commit and also it doesn't bring any benefit since we squash them when merging.

joeldenning commented:
@arturovt do you have any other comments on this before we merge?

KevinCathcart commented:
Ok cool. I'd be happy to create a PR that cherry picks this into 3.x this
evening. Having already tried that, I know it is not quite a clean merge,
but there is no significant difficulty there. (Biggest issue is some of the
files moved in a way that git failed to detect. Also a a few small linting
and lack of type annotation differences.)

On Thu, Jul 2, 2020, 11:58 Joel Denning <notifications@github.com> wrote:

> *@joeldenning* approved this pull request.
>
> This looks good to me. We should also try to put together a fix for the
> 3.x branch for Angular versions 2-8.
>
> Also, I am okay with merging this in before the documentation is authored
> and merged.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/single-spa/single-spa-angular/pull/229#pullrequestreview-441827193>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWEZJA2D5RKORJOKRAVIDRZSVDLANCNFSM4OD7OLNA>
> .
>

joeldenning commented:
I put together the documentation for this at https://github.com/single-spa/single-spa.js.org/pull/293. @KevinCathcart could you review that pull request?

arturovt commented on src/webpack/index.ts:
```diff
@@ -75,3 +85,42 @@ function removeMiniCssExtract(config: any) {
+  const angularJson = JSON.parse(fs.readFileSync(angularJsonPath, 'utf8'));
+  if (!angularJson.projects) return null;
+
+  var projects = Object.keys(angularJson.projects);
```
```suggestion
  const projects = Object.keys(angularJson.projects);
```

joeldenning commented:
👍 I'll merge and publish this tomorrow morning.

joeldenning commented:
This is released in https://github.com/single-spa/single-spa-angular/releases/tag/v4.3.1

@KevinCathcart if you're up to PR'ing the `3.x` branch for Angular versions 2-8, that would be greatly appreciated! Thanks again for your work on this - was a high quality PR.

joeldenning commented:
Oops I just saw https://github.com/single-spa/single-spa-angular/pull/236 where you have already created the PR for the 3.x branch. Thanks!

KevinCathcart commented:
The 3.x version is PR #236

On Sun, Jul 5, 2020, 23:46 Joel Denning <notifications@github.com> wrote:

> This is released in
> https://github.com/single-spa/single-spa-angular/releases/tag/v4.3.1
>
> @KevinCathcart <https://github.com/KevinCathcart> if you're up to PR'ing
> the 3.x branch for Angular versions 2-8, that would be greatly
> appreciated! Thanks again for your work on this - was a high quality PR.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/single-spa/single-spa-angular/pull/229#issuecomment-653999816>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAAWEZP2JUH6TQENDFLOQ73R2FCJ3ANCNFSM4OD7OLNA>
> .
>
